// ngx-pipes/src/app/pipes/helpers/helpers.js
"use strict";
function isUndefined(value) {
    return typeof value === 'undefined';
}
exports.isUndefined = isUndefined;
function isFunction(value) {
    return typeof value === 'function';
}
exports.isFunction = isFunction;
function isNumber(value) {
    return typeof value === 'number';
}
exports.isNumber = isNumber;
function isString(value) {
    return typeof value === 'string';
}
exports.isString = isString;
function isBoolean(value) {
    return typeof value === 'boolean';
}
exports.isBoolean = isBoolean;
function isObject(value) {
    return value !== null && typeof value === 'object';
}
exports.isObject = isObject;
function isNumberFinite(value) {
    return isNumber(value) && isFinite(value);
}
exports.isNumberFinite = isNumberFinite;
function extractDeepPropertyByMapKey(obj, map) {
    var keys = map.split('.');
    var key = keys.shift();
    return keys.reduce(function (prop, key) {
        return !isUndefined(prop) && !isUndefined(prop[key])
            ? prop[key]
            : undefined;
    }, obj[key || '']);
}
exports.extractDeepPropertyByMapKey = extractDeepPropertyByMapKey;
